lib/Requalify.thy

```diff
@@ -1,117 +1,474 @@
+    (gen_requalify get_fact_nm Parse.thm check_fact alias_fact true)
+
+val _ =
+  Outer_Syntax.command @{command_keyword arch_global_naming}
+    "change global naming of context block to \"($L4V_ARCH)_[A|H]\" using env. variable"
+    (arch_global_opts  >> (fn arch_suffix =>
+      Toplevel.local_theory NONE NONE
+        (Local_Theory.map_background_naming
+          (Name_Space.parent_path
+           #> Name_Space.qualified_path true (Binding.make (arch_prefix arch_suffix, @{here}))))))
 
 end
 
 end
 \<close>
 
-(*Tests and examples *)
+section \<open>Tests and examples\<close>
+
+subsection \<open>Generic\<close>
+
+subsubsection \<open>Exporting types, constants and facts from a locale into the theory context\<close>
+
+locale Requalify_Example1
+
+context Requalify_Example1 begin
+typedecl ex1_type
+definition "ex1_const \<equiv> undefined :: ex1_type"
+end
+
+(* these will all generate errors:
+typ ex1_type
+term "ex1_const :: ex1_type"
+thm ex1_const_def
+*)
+
+typ Requalify_Example1.ex1_type
+term "Requalify_Example1.ex1_const :: Requalify_Example1.ex1_type"
+thm Requalify_Example1.ex1_const_def
+
+(* exporting will make types/consts/facts available *)
+
+requalify_types Requalify_Example1.ex1_type
+typ ex1_type
+
+requalify_consts Requalify_Example1.ex1_const
+term "ex1_const :: ex1_type"
+
+requalify_facts Requalify_Example1.ex1_const_def
+thm ex1_const_def
+
+(* trying to export into theory context that already has that name will result in warnings *)
+requalify_types Requalify_Example1.ex1_type
+requalify_consts Requalify_Example1.ex1_const
+requalify_facts Requalify_Example1.ex1_const_def
+
+(* warnings can be suppressed if naming collision is on purpose, but see caveats in next sections *)
+requalify_types (aliasing) Requalify_Example1.ex1_type
+requalify_consts (aliasing) Requalify_Example1.ex1_const
+requalify_facts (aliasing) Requalify_Example1.ex1_const_def
+
+(* requalification can also occur via interpretation, using internal names, but this is slower *)
+context begin interpretation Requalify_Example1 .
+requalify_types ex1_type
+requalify_consts ex1_const
+requalify_facts ex1_const_def
+end
+
+
+subsubsection \<open>Exporting types, constants and facts from a locale into a locale context\<close>
+
+locale Requalify_Example2
+
+(* the target of the export can be a locale (this mode cannot be used from an interpretation) *)
+
+requalify_types (in Requalify_Example2) Requalify_Example1.ex1_type
+requalify_consts (in Requalify_Example2) Requalify_Example1.ex1_const
+requalify_facts (in Requalify_Example2) Requalify_Example1.ex1_const_def
+
+(* this is equivalent to doing the requalifications in the locale context *)
+context Requalify_Example2 begin
+requalify_types (in Requalify_Example2) Requalify_Example1.ex1_type
+requalify_consts (in Requalify_Example2) Requalify_Example1.ex1_const
+requalify_facts (in Requalify_Example2) Requalify_Example1.ex1_const_def
+end
```

Comment:
Just to clarify, since it isn't clear to me from the comment. When requalifying in the locale context you still need to provide the target? On first read of the comment and without looking at the source I would have guessed it to mean the opposite, that providing the target is equivalent to doing it within the context without providing it.

Reply:
Very well spotted, this doesn't make sense due to copy-pasting and a poor comment. I think the example was supposed to do this:
```
(* this is equivalent to doing the requalifications in the locale context *)
context Requalify_Example1 begin
requalify_types (in Requalify_Example2) ex1_type
requalify_consts (in Requalify_Example2) ex1_const
requalify_facts (in Requalify_Example2) ex1_const_def
end
```
Is that better, or is the comment still misleading?

Reply:
Yep, that works for me! I think on its own I might still be confused by the comment but with the example it makes perfect sense now.

Reply:
Hmm, maybe something like `(* this is equivalent to doing the requalifications in the original locale context *)`? I think part of my confusion is due to expecting `the locale context` to refer to the target that the example is talking about.

Reply:
will do

Reply:
done